File: contract/src/investigation.rs
use std::prelude::v1::*;
use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
use near_sdk::serde::{Deserialize, Serialize};
use near_sdk::{AccountId, json_types::U64, env};
use crate::webhook_mappings::WebhookType;

#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(crate = "near_sdk::serde")]
pub enum InvestigationStatus {
    Pending,
    Processing,
    Completed,
    Failed
}

// Add From implementation for WebhookType
impl From<WebhookType> for InvestigationStatus {
    fn from(webhook_type: WebhookType) -> Self {
        match webhook_type {
            WebhookType::Progress => InvestigationStatus::Processing,
            WebhookType::Completion => InvestigationStatus::Completed,
            WebhookType::Error => InvestigationStatus::Failed,
            _ => InvestigationStatus::Processing,
        }
    }
}

#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, Debug)]
#[serde(crate = "near_sdk::serde")]
pub struct AnalysisSummary {
    pub robust_summary: Option<String>,
    pub short_summary: Option<String>,
    pub transaction_count: u64,
    pub is_bot: bool,
}

#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, Debug)]
#[serde(crate = "near_sdk::serde")]
pub struct FinancialSummary {
    pub total_usd_value: String,
    pub near_balance: String,
    pub defi_value: String,
}

#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, Debug)]
#[serde(crate = "near_sdk::serde")]
pub struct InvestigationMetadata {
    pub case_number: u64,
    pub target_account: AccountId,
    pub requester: AccountId,
    pub investigation_date: U64,
    pub status: InvestigationStatus,
    pub summary: Option<String>,         // Combined summary
    pub transaction_count: u64,
    pub total_value_usd: String,
    pub near_balance: String,            // Keep NEAR balance separate
    pub is_bot: bool,                    // Keep bot detection
    pub last_updated: U64,              // For frontend compatibility
}

impl InvestigationMetadata {
    pub fn new(
        case_number: u64,
        target_account: AccountId,
        requester: AccountId,
    ) -> Self {
        Self {
            case_number,
            target_account,
            requester,
            investigation_date: U64(env::block_timestamp()),
            last_updated: U64(env::block_timestamp()),
            status: InvestigationStatus::Pending,
            summary: None,
            transaction_count: 0,
            total_value_usd: "0".to_string(),
            near_balance: "0".to_string(),
            is_bot: false,
        }
    }

    pub fn update_from_webhook(&mut self, webhook_data: &serde_json::Value) {
        // Update status
        if let Some(status) = webhook_data.get("status").and_then(|v| v.as_str()) {
            self.status = match status {
                "complete" => InvestigationStatus::Completed,
                "processing" => InvestigationStatus::Processing,
                "failed" => InvestigationStatus::Failed,
                _ => self.status.clone(),
            };
        }

        if let Some(result) = webhook_data.get("result") {
            // Update financial data
            if let Some(financial) = result.get("financialData") {
                if let Some(total_usd) = financial.get("totalUsdValue").and_then(|v| v.as_str()) {
                    self.total_value_usd = total_usd.to_string();
                }
                if let Some(near_balance) = financial.get("nearBalance").and_then(|v| v.as_str()) {
                    self.near_balance = near_balance.to_string();
                }
            }

            // Update analysis data
            if let Some(robust) = result.get("robustSummary").and_then(|v| v.as_str()) {
                self.summary = Some(robust.to_string());
            } else if let Some(short) = result.get("shortSummary").and_then(|v| v.as_str()) {
                self.summary = Some(short.to_string());
            }
            if let Some(count) = result.get("transactionCount").and_then(|v| v.as_u64()) {
                self.transaction_count = count;
            }
            if let Some(is_bot) = result.get("isBot").and_then(|v| v.as_bool()) {
                self.is_bot = is_bot;
            }
        }

        self.last_updated = U64(env::block_timestamp());
    }
}

File: src/types/investigation.ts
// Import WebhookType from webhook.ts
import { WebhookType } from './webhook';

// Define valid status types
export type InvestigationStatus = 'Pending' | 'Processing' | 'Completed' | 'Failed';

// Define timestamp type to match contract's U64
export type NearTimestamp = string | number;

// Core investigation data (stored in 'extra' field of NFT)
export interface InvestigationMetadata {
    case_number: number;
    target_account: string;
    requester: string;
    investigation_date: NearTimestamp;
    last_updated: NearTimestamp;
    status: InvestigationStatus;
    financial_summary: {
        total_usd_value: string;
        near_balance: string;
        defi_value: string;
    };
    analysis_summary: {
        robust_summary: string | null;
        short_summary: string | null;
        transaction_count: number;
        is_bot: boolean;
    };
}

// Account metadata from external sources
export interface AccountMetadata {
    // Core account info
    accountId: string;
    created: string;
    last_updated: string;
    tx_count: number;

    // Financial data
    wealth: {
        defi: {
            totalUSDValue: string;
        };
        balance: {
            items: {
                contract: string;
                amount: number;
                symbol: string;
                usdValue: string;
                tokenPrice?: string;
            }[];
            totalUSDValue: string;
        };
        totalUSDValue: string;
    };

    // Analysis data
    bot_detection: {
        isPotentialBot: boolean;
    };
    
    // Status tracking
    status: 'completed' | 'processing' | 'failed';
    taskId: string;
}

// Investigation summaries interface
export interface InvestigationSummaries {
    robustSummary: string | null;
    shortSummary: string | null;
}

// The actual NFT metadata structure
export interface InvestigationNFTMetadata {
    title: string;          // Displayed as NFT title
    description: string;    // Displayed as NFT description
    media: string;         // NFT image URL
    media_hash?: string | null;
    copies?: number;
    issued_at: string;
    extra: InvestigationMetadata;
}

// Webhook data structure (moved from webhook.ts if you want to consolidate)
export interface WebhookData {
    type: WebhookType;
    status: 'processing' | 'complete' | 'failed';
    data: {
        accountId: string;
        taskId: string;
        requestId: string;
        progress?: number;
        message?: string;
        result?: {
            robustSummary: string;
            shortSummary: string;
            transactionCount: number;
            isBot: boolean;
            financialData: {
                totalUsdValue: string;
                nearBalance: string;
                defiValue: string;
            }
        };
        error?: string;
        timestamp: string;
    };
}

// Updated helper function to ensure consistent metadata structure
export function formatInvestigationMetadata(
    accountMetadata: AccountMetadata,
    summaries: InvestigationSummaries,
    caseNumber: number = 0,  // You'll need to pass this from the contract
    requester: string = ''   // You'll need to pass this from the contract
): InvestigationNFTMetadata {
    return {
        title: `Case File #${caseNumber}: ${accountMetadata.accountId}`,
        description: summaries.robustSummary || "Investigation in progress...",
        media: process.env.NEXT_PUBLIC_DEFAULT_NFT_IMAGE_URL!,
        media_hash: null,
        copies: 1,
        issued_at: new Date().toISOString(),
        extra: {
            case_number: caseNumber,
            target_account: accountMetadata.accountId,
            requester: requester,
            investigation_date: accountMetadata.created,
            last_updated: accountMetadata.last_updated,
            status: accountMetadata.status === 'completed' ? 'Completed' 
                 : accountMetadata.status === 'processing' ? 'Processing'
                 : accountMetadata.status === 'failed' ? 'Failed'
                 : 'Pending',
            financial_summary: {
                total_usd_value: accountMetadata.wealth.totalUSDValue,
                near_balance: accountMetadata.wealth.balance.items
                    .find(item => item.symbol === "NEAR")?.amount.toString() || "0",
                defi_value: accountMetadata.wealth.defi.totalUSDValue
            },
            analysis_summary: {
                robust_summary: summaries.robustSummary,
                short_summary: summaries.shortSummary,
                transaction_count: accountMetadata.tx_count,
                is_bot: accountMetadata.bot_detection.isPotentialBot
            }
        }
    };
}